### DISCUSSION TITLE
Allowing for other languages in Quarto

### DISCUSSION THREAD

**Discussion by friosavila** [quarto-dev/quarto-cli#4757] [Q&A] [answered]
Hi, 
I was wondering if there is any guide on how to add functionality to other languages when using quarto.
Say, for example, I developed a kernel call, icoolprog for my language coolprog.
I can use this in jupyter, but when trying to use it in Rstudio or VScode (with quarto extension) the functionality that allows to "run" a code block is absent. 
Is there any way to adjust settings in Quatro to allow for that?
Thank you

**Comment by cderv** [comment]
> Say, for example, I developed a kernel call, icoolprog for my language coolprog.

Quarto allows you to adjust kernel https://quarto.org/docs/computations/python.html#kernel-selection
So if it works with Jupyter, it should work with Quarto

I believe you have tried that already, so it may need some adjustment. Can you share example of what you tried ? 

Or is it just a theoretical question ? 

Regarding multilanguage, document with R use **knitr** which allows to run[ multiple type of language](https://bookdown.org/yihui/rmarkdown-cookbook/other-languages.html) too (but it will require R so not always desirable)

**Reply by cderv** [reply]
Sorry I misunderstood the question. You are talking about interactive cell execution in IDE. I'll hide the above answer

**Comment by hugetim** [comment]
For VSCode, specifically, it says [here](https://github.com/quarto-dev/quarto/tree/main/apps/vscode#code-cells) that only Python, Julia, and R cell execution is supported.

**Comment by mcanouil** [comment, marked as answer]
To add languages, it's either:
- add an engine to `knitr`, see <https://bookdown.org/yihui/rmarkdown-cookbook/custom-engine.html>
- develop/use your own kernel and set it in Quarto, see <https://quarto.org/docs/computations/python.html#kernel-selection>

I am going to switch this to "Q&A".

**Reply by hugetim** [reply]
This does not answer the specific request about how to enable "Run Cell" in VS Code, and it's confusing to have it marked as the answer.

**Comment by lukts30** [comment]
So I got quarto extension set up in vscode and command palette "Quarto: Preview" does work with both Rust ([evcxr](https://github.com/evcxr/evcxr)) and Deno. Code written in the these block does render in the created HTML. 
Code completion and syntax highlighting also work, but I do not have the "Run Cell" inline hint, nor does the "Quarto: Run Cell" command palette do anything.

The Microsoft Jupyter extension does work with both these kernels without any issue.

````
```{rust}
```{typescript}
````
````
---
title: "Quarto Deno Test"
jupyter: deno
---

Hello world test!

```{typescript}
console.log("Hello World!!!")
```
````

These languages show the inline hint even though only python is installed and working. 
````
```{python}
```{julia}
```{r}
````

What it the reason for [hardcoding](https://github.com/quarto-dev/quarto/tree/main/apps/vscode#code-cells) to only python, julia and r ?

**Reply by cderv** [reply]
This would be a feature request to https://github.com/quarto-dev/quarto

> What it the reason for [hardcoding](https://github.com/quarto-dev/quarto/tree/main/apps/vscode#code-cells) to only python, julia and r ?

First because this are all language support by Quarto natively. Rust or Typescript is not a known language to quarto. It goes through one of the engine. So there needs to be some knowledge for the extension that they exist and can be run; 

Though this is also possibly because integration with code execution is done through other extension to run the cells. And evxr or Deno needs to be "taught" to the extension. 

So this is a feature request to open in https://github.com/quarto-dev/quarto

There is one for Rcpp cell for example:
- https://github.com/quarto-dev/quarto/issues/438

Not sure what can be done, especially because there is so many options of interactive available when you consider all available Jupyter Kernel, and custom knitr engine. Should quarto extension know them all to actvate the run cell button ? Maybe there is a way that does not imply to know them all if an other extension can communicate support. Anyhow, let's discuss this in the right repo. Thank you !

**Reply by hugetim** [reply]
It's actually no longer true that it only works for python, julia, and r. Stata was added in https://github.com/quarto-dev/quarto/commit/5e0c7b5af4d23211d9028d31536c9a63b79a977e (see https://github.com/quarto-dev/quarto/issues/336#issuecomment-1873437072). Also supported currently: bash (with "sh" and "shell" as aliases).

**Reply by hugetim** [reply]
I went ahead and created an issue at the appropriate repo: https://github.com/quarto-dev/quarto/issues/597
